### PR TITLE
feat: Time line data

### PR DESCRIPTION
**Pull request by heiyu4585**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

https://github.com/ant-design/ant-design/issues/39265
### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |    time-line 改为数据驱动  |
| 🇨🇳 Chinese |     time-line support `items`   |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/timeline/Timeline.tsx [hunk 1]
```diff
@@ -37,56 +39,101 @@ const Timeline: CompoundedComponent = (props) => {
...
-    const pendingClass = idx === itemsCount - 2 ? lastCls : '';
-    const readyClass = idx === itemsCount - 1 ? lastCls : '';
-    return cloneElement(ele, {
-      className: classNames([
-        ele.props.className,
-        !reverse && !!pending ? pendingClass : readyClass,
-        getPositionCls(ele, idx),
-      ]),
+  //  support items
+  if (hasItems) {
```
Comment (zombieJ) [inline components/timeline/Timeline.tsx:73]: 这一块和 children 处理最好抽成 hooks，然后渲染节点的逻辑额外做一个 component 管理一下做解耦：
```tsx
const mergedItems = useItems(items, children);

return (
  <...>
    <ItemList items={mergedItems} />
  </...>
);
```

Comment (heiyu4585) [reply components/timeline/Timeline.tsx]: done

File: components/timeline/TimelineItemList.tsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      if (position === 'right') return `${prefixCls}-item-right`;
+      if (position === 'left') return `${prefixCls}-item-left`;
+      return idx % 2 === 0 ? `${prefixCls}-item-left` : `${prefixCls}-item-right`;
+    }
+    if (mode === 'left') return `${prefixCls}-item-left`;
+    if (mode === 'right') return `${prefixCls}-item-right`;
+    if (position === 'right') return `${prefixCls}-item-right`;
+    return '';
+  };
+  const mergeItems = (items || []).concat([]);
```
Comment (heiyu4585) [inline components/timeline/TimelineItemList.tsx:33]: 这个地方有点拿不准,是否有更好的写法

Comment (zombieJ) [reply components/timeline/TimelineItemList.tsx:33]: 这个不是深拷贝，不过可以简化一下：
```tsx
const mergedItems = [...(items || [])];
```

File: components/timeline/__tests__/__snapshots__/demo-extend.test.ts.snap [hunk 3]
```diff
@@ -231,21 +231,6 @@ exports[`renders ./components/timeline/demo/color.tsx extend context correctly 1
       Create a services site 2015-09-01
     </div>
   </li>
-  <li
```
Comment (zombieJ) [inline components/timeline/__tests__/__snapshots__/demo-extend.test.ts.snap:234]: snapshot 变了

File: components/timeline/useItems.tsx [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import type { TimelineItemProps } from './TimelineItem';
+
+function useItems(items?: TimelineItemProps[], children?: React.ReactNode): TimelineItemProps[] {
+  if (items && Array.isArray(items)) return items;
+
+  const truthyItems = React.Children.toArray(children).filter((item) => !!item);
```
Comment (zombieJ) [inline components/timeline/useItems.tsx:7]: `rc-util` 有个 toArray 可以用，会自动过滤

Comment (zombieJ) [reply components/timeline/useItems.tsx:7]: 然后直接调用 map 即可，不用再用 `React.Children.map`

File: components/timeline/__tests__/__snapshots__/demo-extend.test.ts.snap (hunk 3 above)
Comment (heiyu4585) [reply components/timeline/__tests__/__snapshots__/demo-extend.test.ts.snap:234]: 这个是特意改的,因为用content做key,会有相同的key warning,所以删掉了

Comment (zombieJ) [reply components/timeline/__tests__/__snapshots__/demo-extend.test.ts.snap:234]: 那不对的，context 相同不应该报 warning，应该和 jsx 一样是有 `key` 用 key，没有则用 index 做 key

Comment (heiyu4585) [reply components/timeline/__tests__/__snapshots__/demo-extend.test.ts.snap:234]: done

File: components/timeline/useItems.tsx (hunk 4 above)
Comment (heiyu4585) [reply components/timeline/useItems.tsx]: done

File: components/timeline/TimelineItemList.tsx (hunk 2 above)
Comment (heiyu4585) [reply components/timeline/TimelineItemList.tsx]: done

**Review by zombieJ — APPROVED** [review]
改一下 5.3.0，其他的没问题了

**Comment by heiyu4585** [comment]
> 改一下 5.3.0，其他的没问题了

done

**Comment by li-jia-nan** [comment]
5.2不是还没发么，为啥不搭个顺风车🚗

**Pull request merged**